signac/contrib/project.py

```diff
@@ -2552,42 +2574,71 @@ def groupby(self, key=None, default=None):
                 else:
                     _filter = {"$and": [{key: {"$exists": True}}, _filter]}
 
-                def keyfunction(job):
-                    """Return job's state point value corresponding to the key.
+                if "." in key and key.split(".", 1)[0] == "doc":
```

Comment:
Okay this next section feels a bit wild, lots of logic to determine what `keyfunction` to use, off the top of my head I can't think of a really easy way to refactor this to make the logic more simple, but could you add a few developer comments to guide the reader what is going on here? Some of the string manipulation happening here is a bit opaque.

Reply:
Oh yeah, I remember hating this part of the merge. Let me see what I can do.

Reply:
Hopefully this logic is clearer now.

Reply:
Amazingly clearer!